### PR TITLE
Code structure rework and unit testing

### PR DESCRIPTION
**Pull request by crmsnwl**

### CODE REVIEW

File: CityPuzzleTests/Classes/RoomTests.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        [Theory]
+        [MemberData(nameof(GetUsers))]
+        public void setParticipantsTestPass(User user)
+        {
+            Room room = new Room() { ID = "", Owner = 0, RoomSize = 0};
+
+            room.setParticipants(user);
+
+            Assert.Contains(1478, room.ParticipantIDs);
```
Comment (AllBill127) [inline CityPuzzleTests/Classes/RoomTests.cs:19]: Should only be one assert

File: CityPuzzleTests/Classes/RoomTests.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            Assert.Contains(1410, room.ParticipantIDs);
```
Comment (AllBill127) [inline CityPuzzleTests/Classes/RoomTests.cs:20]: Use user.ID instead of integer literal to assert if ID waas added

**Comment by crmsnwl** [comment]
Fixed the tests according to comments.

File: CityPuzzle/Login And SingUp/LoginPage.xaml [hunk 3]
```diff
@@ -8,12 +8,12 @@
...
-        <Entry x:Name="Vartotojo_vardas" Placeholder="Vartotojo vardas"  ClearButtonVisibility="WhileEditing" HorizontalOptions="Fill" BackgroundColor="#F3BD21" ></Entry>
+        <Entry x:Name="username" Placeholder="Vartotojo vardas"  ClearButtonVisibility="WhileEditing" HorizontalOptions="Fill" BackgroundColor="#F3BD21" ></Entry>
         
-        <Entry x:Name="Slaptazodis" Placeholder="Slaptazodis" IsPassword="true" BackgroundColor="#F3BD21" ></Entry>
+        <Entry x:Name="password" Placeholder="Slaptažodis" IsPassword="true" BackgroundColor="#F3BD21" ></Entry>
         <Label Padding="0,0,0,50"></Label>
-        <Button x:Name="Login_Click_Button" Text="Prisijungti" Clicked="Login_Click" BackgroundColor="#F3BD21"  Padding="0,0,0,0"></Button>
-        <Button x:Name="Sign_Click_Button" Text="Registruotis" Clicked="Sign_Click" BackgroundColor="#F3BD21"></Button>
+        <Button x:Name="Login_Clicked" Text="Prisijungti" Clicked="Login_Click" BackgroundColor="#F3BD21"  Padding="0,0,0,0"></Button>
+        <Button x:Name="SignIn_Clicked" Text="Registruotis" Clicked="Sign_Click" BackgroundColor="#F3BD21"></Button>
```
Comment (crmsnwl) [inline CityPuzzle/Login And SingUp/LoginPage.xaml:16]: x:Name should be Login_Click and Clicked="Login_Clicked". Same with SignIn.

File: CityPuzzle/Side Pages/AddObjectPage.xaml.cs [hunk 4]
```diff
@@ -17,23 +17,29 @@ public partial class AddObjectPage : ContentPage
...
+                Puzzle obj = new Puzzle()
+                {
+                    About = objectAbout.Text,
+                    Quest = objectQuest.Text,
+                    Name = objectName.Text,
+                    ImgAdress = objectImg.Text,
+                    Latitude = Convert.ToDouble(objectX.Text),
+                    Longitude = Convert.ToDouble(objectY.Text),
+                };
+                Sql.SavePuzzle(obj);
```
Comment (AllBill127) [inline CityPuzzle/Side Pages/AddObjectPage.xaml.cs:39]: Use new Puzzle.Save() method from API

File: CityPuzzle/Classes/AddPageLogic.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            allPuzzles = Sql.ReadPuzzles();
+        }
+
+        public void ChangePuzzle(bool direction)
+        {
+            if (allPuzzles.Count == 0)
+            {
+                OnNoPuzzlesFound?.Invoke(this, EventArgs.Empty);
+            }
+            else if (SelectedPuzzles.Contains(allPuzzles[puzzleNr++]) && direction)
```
Comment (AllBill127) [inline CityPuzzle/Classes/AddPageLogic.cs:33]: using ++ operator in this situation is incorrect leading to unexpected errors. Should be puzzleNr + 1

File: CityPuzzle/Classes/AddPageLogic.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                ChangePuzzle(true);
+            }
+            else if (SelectedPuzzles.Contains(allPuzzles[puzzleNr--]) && !direction)
```
Comment (AllBill127) [inline CityPuzzle/Classes/AddPageLogic.cs:37]: allPuzzles -- and even puzzleNr -1 can lead to indexOutOfBoundsException. Should add extra check before using such indexing

File: CityPuzzle/Classes/AddPageLogic.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                puzzleNr--;
+                OnPuzzleChange?.Invoke(this, new OnPuzzleChangeEventArgs { Name = allPuzzles[puzzleNr].Value.Name, ImgAdress = allPuzzles[puzzleNr].Value.ImgAdress, About = allPuzzles[puzzleNr].Value.About });
+            }
+        }
+
+        public void AddPuzzle()
+        {
+            SelectedPuzzles.Add(allPuzzles[puzzleNr]);
+            ChangePuzzle(true);
```
Comment (AllBill127) [inline CityPuzzle/Classes/AddPageLogic.cs:65]: Going up a list is imposible when last puzzle is added so this should have a case for that situation and go down

File: CityPuzzle/Game Room/Create GameRoom/AddPage.xaml [hunk 8]
```diff
@@ -11,9 +11,10 @@
             <Frame BackgroundColor="#F3BD21" Padding="24" CornerRadius="10">
                 <Label x:Name="PuzzleInfo" HorizontalOptions="CenterAndExpand" FontSize="18" TextColor="Black" ></Label>
             </Frame>
-            <Button Text="Prideti" Clicked="Add_puzzle"></Button>
-            <Button Text="Kitas" Clicked="Next_puzzle"></Button>
-
+            <Button Text="Prideti" Clicked="AddPuzzle_Clicked" />
```
Comment (AllBill127) [inline CityPuzzle/Game Room/Create GameRoom/AddPage.xaml:14]: Lithuanian UI. Check spelling

File: CityPuzzle/Game Room/Create GameRoom/SelectPuzzles.xaml.cs [hunk 9]
```diff
@@ -19,10 +19,10 @@ public SelectPuzzles(List<T> given)
 
             var layout = new StackLayout { Padding = new Thickness(5, 10) };
             Label label = new Label();
-            label.Text="Redaguokite sarasa";
+            label.Text = "Redaguokite sarasa";
```
Comment (AllBill127) [inline CityPuzzle/Game Room/Create GameRoom/SelectPuzzles.xaml.cs:22]: Fix structure of the code by adding some whitespaces for better readability

File: CityPuzzle/Rest Services/Client/HttpClientRequest.cs [hunk 10]
```diff
@@ -13,9 +13,19 @@ public class HttpClientRequest
...
-            return content;
+            Console.WriteLine("url + objectPath" + url + objectPath);
+            Task<string> sendcommand = httpClient.GetStringAsync(url + objectPath);
+            Thread timer = new Thread(new ThreadStart(startTimer));
+            timer.Start();
+            while (timer.IsAlive)
+            {
+                if (sendcommand.IsCompleted)
+                {
+                    return sendcommand.Result;
```
Comment (AllBill127) [inline CityPuzzle/Rest Services/Client/HttpClientRequest.cs:26]: after return the startTimer thread is not destroyed so after 3s it throws exception no matter what

File: CityPuzzle/Rest Services/Client/HttpClientRequest.cs [hunk 11]
```diff
@@ -53,6 +63,9 @@ public static string GetAdress<T>(T item)
                 return "Tasks";
             throw new Classes.TypeNotExistException();
         }
-
+        public void startTimer()
```
Comment (AllBill127) [inline CityPuzzle/Rest Services/Client/HttpClientRequest.cs:70]: could be done using lambda expression while declaring the thread

**Comment by AllBill127** [comment]
Code clean up is done. Please review some changes. We can proceed with testing

**Pull request merged**